internal/controllers/machinedeployment/machinedeployment_sync.go

```diff
@@ -86,173 +89,266 @@ func (r *Reconciler) getAllMachineSetsAndSyncRevision(ctx context.Context, d *cl
+	//   machine template spec (take the one from MachineDeployment)
+	if existingMS == nil {
+		uniqueIdentifierLabelValue = apirand.String(5)
+
+		name = computeNewMachineSetName(deployment.Name+"-", uniqueIdentifierLabelValue)
+
+		// Add foregroundDeletion finalizer to MachineSet if the MachineDeployment has it.
+		if sets.New[string](deployment.Finalizers...).Has(metav1.FinalizerDeleteDependents) {
+			finalizers = []string{metav1.FinalizerDeleteDependents}
+		}
 
-	// Set new machine set's annotation
-	mdutil.SetNewMachineSetAnnotations(d, &newMS, newRevision, false, log)
-	// Create the new MachineSet. If it already exists, then we need to check for possible
-	// hash collisions. If there is any other error, we need to report it in the status of
-	// the Deployment.
-	alreadyExists := false
-	err = r.Client.Create(ctx, &newMS)
-	createdMS := &newMS
-	switch {
-	// We may end up hitting this due to a slow cache or a fast resync of the Deployment.
-	case apierrors.IsAlreadyExists(err):
-		alreadyExists = true
+		replicas, err = mdutil.NewMSNewReplicas(deployment, oldMSs, 0)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to compute desired MachineSet")
+		}
 
-		ms := &clusterv1.MachineSet{}
-		msErr := r.Client.Get(ctx, client.ObjectKey{Namespace: newMS.Namespace, Name: newMS.Name}, ms)
-		if msErr != nil {
-			return nil, msErr
+		machineTemplateSpec = *deployment.Spec.Template.Spec.DeepCopy()
+	} else {
+		// For updating an existing MachineSet:
+		// * get the uniqueIdentifier from labels of the existingMS
+		// * use name, uid, finalizers, replicas and machine template spec from existingMS.
+		// Note: We use the uid, to ensure that the Server-Side-Apply only updates existingMS.
+		// Note: We carry over those fields because we don't want to mutate them for an existingMS.
+		var uniqueIdentifierLabelExists bool
+		uniqueIdentifierLabelValue, uniqueIdentifierLabelExists = existingMS.Labels[clusterv1.MachineDeploymentUniqueLabel]
+		if !uniqueIdentifierLabelExists {
+			return nil, errors.Errorf("failed to compute desired MachineSet: failed to get unique identifier from %q annotation",
+				clusterv1.MachineDeploymentUniqueLabel)
 		}
 
-		// If the Deployment owns the MachineSet and the MachineSet's MachineTemplateSpec is semantically
-		// deep equal to the MachineTemplateSpec of the Deployment, it's the Deployment's new MachineSet.
-		// Otherwise, this is a hash collision and we need to increment the collisionCount field in
-		// the status of the Deployment and requeue to try the creation in the next sync.
-		controllerRef := metav1.GetControllerOf(ms)
-		if controllerRef != nil && controllerRef.UID == d.UID && mdutil.EqualMachineTemplate(&d.Spec.Template, &ms.Spec.Template) {
-			createdMS = ms
-			break
+		name = existingMS.Name
+		uid = existingMS.UID
+
+		// Keep foregroundDeletion finalizer if the existingMS has it.
+		if sets.New[string](existingMS.Finalizers...).Has(metav1.FinalizerDeleteDependents) {
```

Comment:
This seems to be done both for new and existing MS, can we move it out from the if else?

Reply:
It's not the same, but we should add a comment there.

In the create case:
* we want to add the finalizer if at this point in time the finalizer exists in the MachineDeployment

In the update case:
* we want to preserve the finalizer if the current MachineSet already has the finalizer

The second part is something we have to do now because of Server-Side-Apply and if we don't do it later reconciles would remove the finalizer.

I think it's important to have exactly this behavior because otherwise the MD reconciler would remove the finalizer from a MachineSet when the finalizer has been removed from the MD.

Reply:
Added a comment to explain the how the update case is special and different from the create case. (This is the case that needs more explanation, the create case is straight-forward).

Reply:
Sounds good to me!